### PR TITLE
Multiple return points

### PR DESCRIPTION
**Pull request by funivan**

### CODE REVIEW

File: src/test/resources/fixtures/codeStyle/multiple-return-points.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            return false;
+        }
+
+        return true;
+    }
+}
+
+class Test_B
+{
+    public function customMethod()
```
Comment (kalessil) [inline src/test/resources/fixtures/codeStyle/multiple-return-points.php]: What about this example?

```php
private function ...() {
    return function(){ return ''; };
}
```

File: src/main/resources/inspectionDescriptions/MultipleReturnPointsInspection.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<html>
+<body>
+Reports methods with too many return points.<br>
+Methods with too many return points may be confusing, and hard to refactor.
+<br />
+<br />
+Please reference to corresponding <a href="https://stackoverflow.com/questions/36707/should-a-function-have-only-one-return-statement">stackoverflow thread</a>
+for more details.
```
Comment (kalessil) [inline src/main/resources/inspectionDescriptions/MultipleReturnPointsInspection.html]: 👍

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @NotNull
+    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
+        return new PhpElementVisitor() {
+
+            public void visitPhpMethod(Method method) {
+                final PhpClass containingClass = method.getContainingClass();
+                if (containingClass != null && !containingClass.isTrait()) {
+
+                    ReturnPointCountVisitor countVisitor = new ReturnPointCountVisitor(method);
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: `final ReturnPointCountVisitor`

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    method.accept(countVisitor);
+
+                    if (countVisitor.getCount() > 1) {
+                        PsiElement nameIdentifier = method.getNameIdentifier();
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: I'd suggest moving this into very beginning of the method

File: src/test/resources/fixtures/codeStyle/multiple-return-points.php (hunk 1 above)
Comment (funivan) [inline src/test/resources/fixtures/codeStyle/multiple-return-points.php]: You mean method without name ?

Comment (kalessil) [inline src/test/resources/fixtures/codeStyle/multiple-return-points.php]: Almost, a method returning a lambda with a single return statement.
Returns in lambdas should not be counted.

Comment (kalessil) [inline src/test/resources/fixtures/codeStyle/multiple-return-points.php]: Thanks! Dropping the comment.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
+        return new PhpElementVisitor() {
+
+            public void visitPhpMethod(Method method) {
+                final PhpClass containingClass = method.getContainingClass();
+                final PsiElement nameIdentifier = method.getNameIdentifier();
+                if (containingClass != null && nameIdentifier != null && !containingClass.isTrait()) {
+
+                    final ReturnPointCountVisitor countVisitor = new ReturnPointCountVisitor(method);
+                    method.accept(countVisitor);
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: @artspb, requesting for support =) 

What are advantages of using visitors vs PsiTreeUtil search methods+regular foreach loop? Are there any, and why if any?

Comment (artspb) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: `PsiTreeUtil` is good for traversing a tree upwards (to the root) or when you need to work with element's children. When you need to traverse a tree downwards (to the leafs) and you don't know an exact position of an element you're looking for or you need to work with several elements, a recursive element visitor is the best choice. It uses a "regular foreach loop" under the hood, but hides this irrelevant detail which makes the code more descriptive and allows concentrating on what you really want to do.

Working with PHP you usually use `PhpElementVisitor`. There's a recursive version of it called `PhpRecursiveElementVisitor`. It's marked as deprecated to recall you, that you should use it really carefully. When a tree is huge you can easily get a SOE or consume a lot of CPU for nothing. To avoid it a tree traversing should be stopped as soon as the desired element is found. For instance, the current implementation contains this mistake and will check each and every element for `instanceof PhpReturn`. A tree traversing won't be stopped even when more than 1 return point is found. That's why we usually don't recommend to use it at all.

**tl;dr**
1. Use `PsiTreeUtil` to find a parent or nearest children.
2. Use `PhpElementVisitor` to make your code more descriptive.
3. Never ever use `PhpRecursiveElementVisitor` (or any handmade recursive visitor) until you understand what you do. Don't forget to profile your code on really deep trees.

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: That a great explanation, thank you!

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    @NotNull
+    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
+        return new PhpElementVisitor() {
+
+            public void visitPhpMethod(Method method) {
+                final PhpClass containingClass = method.getContainingClass();
+                final PsiElement nameIdentifier = method.getNameIdentifier();
+                if (containingClass != null && nameIdentifier != null && !containingClass.isTrait()) {
+
+                    final ReturnPointCountVisitor countVisitor = new ReturnPointCountVisitor(method);
```
Comment (artspb) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: As I mentioned, that a recursive visitor must be avoided, I should provide an alternative. In the context of this task, I'd recommend using a control flow. It can be easily obtained calling `method.getControlFlow()`. A building of a control flow is a kinda expensive task, but once built it's cached and can be reused by several inspections. So most probably some other inspection is already built it for you and you'll get it for free. (Disclaimer: never trust such sentences and always re-check using a profiler.)

Here's the sketch how the code can look like.

    final PhpControlFlow controlFlow = method.getControlFlow();
    final PhpExitPointInstruction exitPoint = controlFlow.getExitPoint();
    int count = 0;
    for (final PhpInstruction instruction : exitPoint.getPredecessors()) {
      if (instruction instanceof PhpReturnInstruction) {
        count++;
      }
      if (count > 1) {
        holder.registerProblem(...);
        return;
      }
    }

Most probably this code doesn't cover some corner cases like `finally`. So you should check how does it work and adjust accordingly. The `PhpControlFlowUtil` class can be helpful. (Note: it's not a part of the open API thus a subject to change.)

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: That's really neat, how long the control flow stuff being there?

Comment (artspb) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: Since forever :)

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: Oh my, I have to learn again. Both funny and annoying %)

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: Thank you @artspb !!!

Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: @kalessil should I rewrite code with using control flow technique ?

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: @funivan : yes please, I'll take care about profiling the final version.

Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: @kalessil done =)

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/MultipleReturnPointsInspector.java]: 👍, add please also @artspb in co-authors please =)

**Comment by kalessil** [comment]
Thank @funivan! Evaluating performance topic this week.
